core/templates/services/messenger.service.ts

```diff
@@ -19,147 +19,201 @@
-      },
-      explorationReset: function(payload) {
-        return Boolean(payload.stateName);
-      },
-      explorationCompleted: function() {
-        return true;
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+
+import { LoggerService } from 'services/contextual/logger.service';
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+interface MessageValidatorsType {
+  heightChange: ((payload: {height: number; scroll: number}) => boolean);
+  explorationLoaded: (() => boolean);
+  stateTransition: (
+    (payload: {oldStateName: string; newStateName: string}) => boolean);
+  explorationReset: ((payload: {stateName: string}) => boolean);
+  explorationCompleted: (() => boolean);
+}
+
+interface HeightChangeData {
+  height: number;
+  scroll: number;
+}
+
+interface ExplorationLoadedData {
+  explorationVersion: number;
+  explorationTitle: string;
+}
+
+interface StateTransitionData {
+  explorationVersion: number;
+  oldStateName: string;
+  jsonAnswer: string;
+  newStateName: string;
+}
+
+interface ExplorationCompletedData {
+  explorationVersion: number;
+}
+
+interface GetPayloadType {
+  heightChange: ((data: HeightChangeData) => HeightChangeData);
+  explorationLoaded: ((data: ExplorationLoadedData) => ExplorationLoadedData);
+  stateTransition: ((data: StateTransitionData) => StateTransitionData);
+  explorationCompleted: (
+    (data: ExplorationCompletedData) => ExplorationCompletedData);
+  explorationReset: ((data: string) => {stateName: string});
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class MessengerService {
+  constructor(
+    private loggerService: LoggerService, private windowRef: WindowRef) {}
+
+  HEIGHT_CHANGE: string = 'heightChange';
+  EXPLORATION_LOADED: string = 'explorationLoaded';
+  STATE_TRANSITION: string = 'stateTransition';
+  EXPLORATION_RESET: string = 'explorationReset';
+  EXPLORATION_COMPLETED: string = 'explorationCompleted';
+
+  SUPPORTED_HASHDICT_VERSIONS: string[] = [
+    '0.0.0', '0.0.1', '0.0.2', '0.0.3'
+  ];
+
+  MESSAGE_VALIDATORS: MessageValidatorsType = {
+    heightChange(payload): boolean {
+      var isPositiveInteger = (n: Object): boolean => {
+        return (typeof n === 'number' && n % 1 === 0 && n > 0);
+      };
+      var isBoolean = (b: Object): boolean => {
+        return typeof b === 'boolean';
+      };
+      return isPositiveInteger(payload.height) && isBoolean(
+        payload.scroll);
+    },
+    explorationLoaded(): boolean {
+      return true;
+    },
+    stateTransition(payload): boolean {
+      return Boolean(payload.oldStateName) || Boolean(payload.newStateName);
+    },
+    explorationReset(payload): boolean {
+      return Boolean(payload.stateName);
+    },
+    explorationCompleted(): boolean {
+      return true;
+    }
+  };
+
+  getPayload: GetPayloadType = {
+    heightChange(data) {
```

Comment:
Please add type-information using the interfaces you've defined at top of file.

Reply:
Done.